pkg/reconciler/delivery/modify.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package delivery
+
+import (
+	"time"
+
+	"knative.dev/pkg/ptr"
+	v1 "knative.dev/serving/pkg/apis/serving/v1"
+	listers "knative.dev/serving/pkg/client/listers/serving/v1"
+)
+
+/**************************************************************************************************************** 
+   modifyRouteSpec assigns traffic to all active Revisions according to the algorithm on page 8 of go/mydesigndoc
+   arguments:
+   - route: the current Route object
+   - r: a lister to query the Revisions by name
+   - newRevName: name string of the latest ready Revision
+   - policy: pointer to the Policy struct that commands the rollout process
+   return values:
+   - 1st value: a new route object whose spec field has been written with the desired state
+   - 2nd value: error if anything goes wrong
+****************************************************************************************************************/ 
+func modifyRouteSpec(route *v1.Route, r listers.RevisionLister, newRevName string, policy *Policy) (*v1.Route, error) {
+	// assumption 1: the current Route Status traffic % are all non-zero (any zero entries would not have been written)
+	// assumption 2: the current Route Status traffic entries are ordered from oldest to newest Revision
+	// first identify whether or not newRevName is already in the pool
+	nameListed := false
+	for _, t := range route.Status.Traffic {
+		if t.RevisionName == newRevName {
+			nameListed = true
+			break
+		}
+	}
+
+	// make a slice container to hold the new traffic assignments, and an ordered, lightweight roster of the pool
+	// that contains all current Revision names, INCLUDING the newest one
+	ln := len(route.Status.Traffic)
+	if !nameListed {
+		ln = ln + 1
+	}
+	if ln == 1 {
+		route.Spec.Traffic = []v1.TrafficTarget{{
+			ConfigurationName: route.Name, // assuming namespace/name matches for Route & Config
```

Comment:
you can't though assume that.

Reply:
OK, I will fix this. Should be a relatively easy fix.

Reply:
I reimplemented this part. I still had to assume that route and configuration have the same namespace, although they don't necessarily need to have the same name. Do you think this is a reasonable assumption?